satellite/satellitedb/orders.go

```diff
@@ -214,3 +218,191 @@ func (db *ordersDB) UnuseSerialNumber(ctx context.Context, serialNumber storj.Se
+	// processed orders are rejected
+	for _, request := range requests {
+		stmt, err := db.buildInsertUseSerialStatement(request.OrderLimit)
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(stmt)
+		if err != nil {
+			if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
+				reject(request.OrderLimit.SerialNumber)
+			} else {
+				return nil, Error.Wrap(err)
+			}
+		}
+	}
+
+	// call to get all the bucket IDs
+	query := db.buildGetBucketIdsQuery(len(requests))
+	statement := db.db.Rebind(query)
+
+	args := make([]interface{}, 0, len(requests))
+	for _, request := range requests {
+		args = append(args, request.OrderLimit.SerialNumber.Bytes())
+	}
+
+	rows, err := tx.Query(statement, args...)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	bucketMap := make(map[storj.SerialNumber][]byte)
+	for rows.Next() {
+		var serialNumber, bucketID []byte
+		err := rows.Scan(&serialNumber, &bucketID)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		sn, err := storj.SerialNumberFromBytes(serialNumber)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		bucketMap[sn] = bucketID
+	}
+
+	// build all the bandwidth updates into one sql statement
+	var updateRollupStatement string
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if !rejected {
+			bucketID, ok := bucketMap[request.OrderLimit.SerialNumber]
+			if ok {
+				projectID, bucketName, err := orders.SplitBucketID(bucketID)
+				if err != nil {
+					return nil, errs.Wrap(err)
+				}
+
+				stmt, err := db.buildUpdateBucketBandwidthRollupStatements(request.OrderLimit, request.Order, projectID[:], bucketName, intervalStart)
+				if err != nil {
+					return nil, err
+				}
+				updateRollupStatement += stmt
+
+				stmt, err = db.buildUpdateStorageNodeBandwidthRollupStatements(request.OrderLimit, request.Order, intervalStart)
+				if err != nil {
+					return nil, err
+				}
+				updateRollupStatement += stmt
+			} else {
+				reject(request.OrderLimit.SerialNumber)
+			}
+		}
+	}
+
+	_, err = tx.Exec(updateRollupStatement)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if !rejected {
+			r := &pb.SettlementResponse{
+				SerialNumber: request.OrderLimit.SerialNumber,
+				Status:       pb.SettlementResponse_ACCEPTED,
+			}
+
+			responses = append(responses, r)
+		}
+	}
+	return responses, nil
+}
+
+func (db *ordersDB) buildInsertUseSerialStatement(orderLimit *pb.OrderLimit) (string, error) {
+	hexNodeID, err := db.toHex(orderLimit.StorageNodeId.Bytes())
+	if err != nil {
+		return "", err
+	}
+	hexSN, err := db.toHex(orderLimit.SerialNumber.Bytes())
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("INSERT INTO  used_serials (serial_number_id, storage_node_id) SELECT id, %v FROM serial_numbers WHERE serial_number = %s;\n",
```

Comment:
In general, it's best to avoid using `fmt.Sprintf` when preparing sql statements, because you risk sql injection and other problems. If you use the `db.Prepare` method, and then use `*sql.Stmt` structs to apply args to the query, you avoid these problems.

Maybe that's not possible given the way you're doing it here? But in general I believe it is best practice to avoid string formatting on sql queries.

Reply:
Updated. Left the others per our conversation